src/VisualStudio/Xaml/Impl/Implementation/XamlProjectService.cs

```diff
@@ -83,54 +84,70 @@ public void TrackOpenDocument(string filePath)
-                _xamlProjects.Add(hierarchy, project);
-            }
-
-            if (!project.ContainsSourceFile(filePath))
-            {
-                project.AddSourceFile(filePath);
+                    if (!hierarchy.TryGetName(out var name))
+                    {
+                        return;
+                    }
+
+                    if (!hierarchy.TryGetGuidProperty(__VSHPROPID.VSHPROPID_ProjectIDGuid, out var projectGuid))
+                    {
+                        return;
+                    }
+
+                    var projectInfo = new VisualStudioProjectCreationInfo
+                    {
+                        Hierarchy = hierarchy,
+                        FilePath = hierarchy.TryGetProjectFilePath(),
+                        ProjectGuid = projectGuid
+                    };
+
+                    project = _visualStudioProjectFactory.CreateAndAddToWorkspace(name, StringConstants.XamlLanguageName, projectInfo);
+                    _xamlProjects.Add(hierarchy, project);
+                }
 
-                var documentId = _workspace.CurrentSolution.GetDocumentIdsWithFilePath(filePath).First(d => d.ProjectId == project.Id);
-                var document = _workspace.CurrentSolution.GetDocument(documentId)!;
-                var hasText = document.TryGetText(out var text);
-                if (!hasText || text?.Container.TryGetTextBuffer() == null)
+                if (!project.ContainsSourceFile(filePath))
                 {
-                    var docInfo = _rdt.GetDocumentInfo(docCookie);
-                    var textBuffer = _editorAdaptersFactory.GetDocumentBuffer(docInfo.DocData as IVsTextBuffer);
-                    var textContainer = textBuffer.AsTextContainer();
-                    _workspace.OnDocumentTextChanged(documentId, textContainer.CurrentText, PreservationMode.PreserveIdentity);
+                    project.AddSourceFile(filePath);
+
+                    var documentId = _workspace.CurrentSolution.GetDocumentIdsWithFilePath(filePath).Single(d => d.ProjectId == project.Id);
+                    var document = _workspace.CurrentSolution.GetDocument(documentId)!;
+                    var hasText = document.TryGetText(out var text);
+                    if (!hasText || text?.Container.TryGetTextBuffer() == null)
+                    {
+                        var docInfo = _rdt.GetDocumentInfo(docCookie);
+                        var textBuffer = _editorAdaptersFactory.GetDocumentBuffer(docInfo.DocData as IVsTextBuffer);
+                        var textContainer = textBuffer.AsTextContainer();
+                        _workspace.OnDocumentTextChanged(documentId, textContainer.CurrentText, PreservationMode.PreserveIdentity);
+                    }
+
+                    _rdtDocumentIds[docCookie] = documentId;
                 }
             }
-
-            if (!_rdtEventsCookie.HasValue)
+            catch (ArgumentException)
```

Comment:
Should this be wrapping a more specific call above? Otherwise there's a huge amount of potential places this could throw (some of which may corrupt state.)

Reply:
Good point. This was for rdt.FindDocument() which may not even need it anymore now that I check for the empty path before calling this. I left it to be safe but I agree that just wrapping FindDocument would be safer.